providers/providers.go

```diff
@@ -148,6 +163,28 @@ func newProviderDataFromConfig(providerConfig options.Provider) (*ProviderData,
 	return p, nil
 }
 
+func stringInSlice(element string, list []string) bool {
+	for _, x := range list {
+		if x == element {
+			return true
+		}
+	}
+	return false
+}
+
+// Pick the most appropriate code challenge method for PKCE
+func ParseCodeChallengeMethod(o *options.Options) string {
```

Comment:
It's not generally good practice to pass the whole options structure around, perhaps the `ForceCodeChallengeMethod` should be a part of the `Provider` struct?